python/python/lance/tf/data.py

```diff
@@ -264,6 +264,67 @@ def lance_fragments(dataset: Union[str, Path, LanceDataset]) -> tf.data.Dataset:
         [f.fragment_id for f in dataset.get_fragments()]
     )
 
+def from_lance_take(
+    dataset: Union[str, Path, LanceDataset],
+    indices: tf.data.Dataset,
+    *,
+    columns: Optional[List[str]] = None,
+    batch_size: Optional[int] = None,
+    output_signature: Optional[Dict[str, tf.TypeSpec]] = None,
+) -> tf.data.Dataset:
+    """Create a ``tf.data.Dataset`` from a Lance dataset and take indices.
+    
+    """
+    if not isinstance(dataset, LanceDataset):
+        dataset = lance.dataset(dataset)
+
+    if output_signature is None:
+        schema = dataset.schema
+        output_signature = schema_to_spec(schema)
+    logging.debug("Output signature: %s", output_signature)
+
+    num_rows = dataset.count_rows()
+
+    def take_rows(indices):
+        for indices in indices:
+            if indices.shape == (0,) and batch_size is not None:
+                end = min(indices + batch_size, num_rows)
+                indices = list(range(indices, end))
+            else:
+                indices = indices.tolist()
+            batch = dataset.take(indices, columns=columns)
+            yield {
+                name: column_to_tensor(batch[name], output_signature[name])
+                for name in batch.schema.names
+            }
+    
+    # return tf.data.Dataset.from_generator(
+    #     take_rows,
+    #     output_signature=output_signature,
+    #     args=(indices.as_numpy_iterator(),),
+    # )
+    
+    # Map doesn't seem to like returning a dict, so we flatten as a list and
+    # reconstruct the dict later.
+    output_values = list(val.dtype for val in output_signature.values())
+
+    result = indices.map(
+        lambda y: tf.numpy_function(
+            take_rows,
+            inp=[y],
+            Tout=output_values
+        ),
+        num_parallel_calls=tf.data.experimental.AUTOTUNE,
+    )
+
+    def consolidate(*args):
+        return {
+            name: value
+            for name, value in zip(output_signature.keys(), args)
+        }
+    
+    return result.map(consolidate)
```

Comment:
I'm totally at a loss for how to write this part.

`tf.data.Dataset.from_generator()` can't handle a dataset as input.

`Dataset.map()` doesn't seem to want to return a dictionary of tensors.

And I'm getting a lot of errors I don't understand.